README.md
# Note for Posts and WordPress Plugin Boilerplate

Adding note to Wordpress post or any post types such as page, attachment, woocommerce products, easy digital downloads and other registered Wordpress Post Type.

Employs the best coding pratices for you to create your own plugin by providing you a quick TODOs to make an easier modification and customization.

## Quick TODOs
Just find and replace text using your favorite editor such as sublime, notepad++, text wrangler, etc
> Only Replace If You Are Creating Your Own Plugin
 * @todo replace Plugin Name
 * @todo replace Plugin URI
 * @todo replace Description
 * @todo replace Version
 * @todo replace Text Domain
 * @todo replace Author
 * @todo replace Author URI

Find and Replace to each files.
 * @todo `Note_For_Posts` Main Class Name. Find all and replace text
 * @todo `N4P` - Prefices for Classes. Find all and replace text
 * @todo `n4p` - Prefixes for Functions. Find all and replace text
 * @todo `note` - Post Type. Find all and replace text
 * @todo `NOTE_FOR_POSTS` - Constant. Find all and replace text

Don't forget to change the filename to your own custom plugin.

## Screenshots
### Screenshot Taken from the Widget
![alt tag](https://raw.githubusercontent.com/esstat17/Note-For-Posts/master/note-for-posts/assets/images/screenshot-1.png)

### Screenshot Taken from the Admin
![alt tag](https://raw.githubusercontent.com/esstat17/Note-For-Posts/master/note-for-posts/assets/images/screenshot-3.png)

## Contents

The Note for Posts Plugin includes the following files:

* `.gitignore`. Used to exclude certain files from the repository.
* `CHANGELOG.md`. The list of changes to the core project.
* `README.md`. The file that you’re currently reading.
* A `note-for-posts` directory that contains the source code - a fully executable WordPress plugin.

## Features
* Able to add custom notes (testimonials, feedbacks, comments, etc) to a particular post type such as page, post, product (Woocommerce), download (Easy Digital Downloads), and other registered post type.
* Able to display results through widget and shortcode.
* The Boilerplate is based on the [Plugin API](http://codex.wordpress.org/Plugin_API), [Coding Standards](http://codex.wordpress.org/WordPress_Coding_Standards), and [Documentation Standards](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/).
* All classes, functions, and variables are well-documented so that you know what you need to be changed.
* The Boilerplate uses a strict file organization scheme that correspond both to the WordPress Plugin Repository structure, and that make it easy to organize the files that compose the plugin.
* The project includes a `.pot` file as a starting point for internationalization.

## Installation

Note that this will activate the source code of the Boilerplate, with real functionalities such as meta boxes, custom post type, shortcodes, taxonomies, admin settings.

## Recommended Tools

### i18n Tools

The WordPress Plugin Boilerplate uses a variable to store the text domain used when internationalizing strings throughout the Boilerplate. To take advantage of this method, there are tools that are recommended for providing correct, translatable files:

* [Poedit](http://www.poedit.net/)
* [makepot](http://i18n.svn.wordpress.org/tools/trunk/)
* [i18n](https://github.com/grappler/i18n)

Any of the above tools should provide you with the proper tooling to internationalize the plugin.

## License

The WordPress Plugin Boilerplate is licensed under the GPL v2 or later.

> This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License, version 2, as published by the Free Software Foundation.

> This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

> You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA

A copy of the license is included in the root of the plugin’s directory. The file is named `LICENSE`.

## Important Notes

### Licensing

The WordPress Plugin Boilerplate is licensed under the GPL v2 or later; however, if you opt to use third-party code that is not compatible with v2, then you may need to switch to using code that is GPL v3 compatible.

For reference, [here's a discussion](http://make.wordpress.org/themes/2013/03/04/licensing-note-apache-and-gpl/) that covers the Apache 2.0 License used by [Bootstrap](http://twitter.github.io/bootstrap/).

### Includes

Note that if you include your own classes, or third-party libraries, there are three locations in which said files may go:

* `note-for-posts/includes` is where functionality shared between the admin area and the public-facing parts of the site reside
* `note-for-posts/includes/admin` is for all admin-specific functionality
* `note-for-posts/language` is for languages


# Credits
Official Wordpress Documentation
Wordpress Functions was created by [Elvin D.](https://twitter.com/esstat17).
[WordPress Plugin Boilerplate](https://github.com/DevinVinson/WordPress-Plugin-Boilerplate) - A standardized, organized, object-oriented foundation for building high-quality WP Plugins
[EDD Extension Boilerplate](https://github.com/easydigitaldownloads/EDD-Extension-Boilerplate) serves as a foundation and aims to provide a standardized guide for building extensions. 

## Documentation, FAQs, and More
If you’re interested in writing any documentation or creating tutorials please [let me know](http://innovedesigns.com/contact/)


CHANGELOG.md
# Changelog
* (16 March 2016). First Launch.



note-for-posts/note-for-posts.php
<?php
/**
 * Plugin Name:     Add Note for WP Posts
 * Plugin URI:      http://innovedesigns.com/
 * Description:     Add Notes for WP Posts and a boilerplate for fast WP Plugin Development
 * Version:         1.0.0
 * Author:          esstat17
 * Author URI:      http://innovedesigns.com/
 * Text Domain:     n4p-txt
 *
 *
 * @package         N4P
 * @author          esstat17
 * @copyright       Copyright (c) 2016
 *
 * @todo Replace only if your creating your own Plugin
 * @todo replace Plugin Name
 * @todo replace Plugin URI
 * @todo replace Description
 * @todo replace Version
 * @todo replace Text Domain
 * @todo replace Author
 * @todo replace Author URI
 *
 * @todo Note_For_Posts - Find all and replace text
 * @todo N4P - Find all and replace text
 * @todo n4p - Find all and replace text
 * @todo NOTE_FOR_POSTS - Find all and replace text
 *
 *
 * IMPORTANT! Ensure that you make the following adjustments
 * before releasing your extension:
 *
 * Copyright 2016 (email : esstat17 at GMAIL.com)
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

// Exit if accessed directly
if( !defined( 'ABSPATH' ) ) exit;

/**
 * The core plugin class.
 *
 * This is used to define internationalization, admin-specific hooks, and
 * public-facing site hooks.
 *
 * Also maintains the unique identifier of this plugin as well as the current
 * version of the plugin.
 *
 * @since      1.0.0
 * @package    Note_For_Posts
 * @subpackage Note_For_Posts/includes
 */
if( !class_exists( 'Note_For_Posts' ) ) {

    /**
     * Main Note_For_Posts class
     *
     * @since       1.0.0
     */
    class Note_For_Posts {


        /**
         * @var         object $instance self assigning 
         * @since       1.0.0
         */
        private static $instance; // 

        /**
         * HTML Element Helper Object.
         *
         * @var object
         * @since 1.0.0
         */
        public $html;

        /**
         * Relational DB SELECT query using Native WP DB System.
         *
         * @var object
         * @since 1.0.0
         */
        public $db;

        /**
         * The unique identifier of this plugin.
         *
         * @since    1.0.0
         * @access   protected
         * @var      string    $plugin_name    The string used to uniquely identify this plugin.
         */
        private $plugin_name;

        /**
         * The Class Constructor
         */

        public function __construct() {
            $this->plugin_name = 'note-for-posts'; // @todo replace it with your plugin name
            $this->setup_constants();
            $this->includes();    
            do_action( 'n4p_hooks' ); // This function invokes all functions attached to `ywp_hooks` action hook 
        }

        /**
         * Get active instance
         *
         * @access      public
         * @since       1.0.0
         * @return      object self::$instance The one true NoteForPost
         */
        public static function instance() {
            if( !self::$instance ) {
                self::$instance = new self(); // // assigning the new NoteForPost Class Instance
                self::$instance->db = new N4P_DB_Query();
                self::$instance->html = new N4P_Friendly_HTML();
            }
            return self::$instance;
        }

        /**
         * Define constant if not already set.
         *
         * @param  string $name
         * @param  string|bool $value
         */
        private function define( $name, $value ) {
            if ( ! defined( $name ) ) {
                define( $name, $value );
            }
        }

        /**
         * Setup plugin constants
         *
         * @access      private
         * @since       1.0.0
         * @return      void
         */
        private function setup_constants() {
            // Plugin version
            $this->define( 'NOTE_FOR_POSTS_VER', '1.0.0' );
            // Plugin path
            $this->define( 'NOTE_FOR_POSTS_DIR', plugin_dir_path( __FILE__ ) );
            // Plugin URL
            $this->define( 'NOTE_FOR_POSTS_URL', plugin_dir_url( __FILE__ ) );

            // Attaching to Post Type: post (default)
            // Getting the value from get_option() or `post` as default
            $post_type = !get_option('n4p_option_1') ? 'post' : get_option('n4p_option_1'); 
            $this->define( 'NOTE_FOR_POSTS_PTYPE', $post_type);
        }
        
        /**
         * Run action and filter hooks
         *
         * @access      private
         * @since       1.0.0
         * @return      void
         *
         * @todo        The hooks listed in this section are a guideline, and
         *              may or may not be relevant to your particular plugin.
         *              Please remove any unnecessary lines, and refer to the
         *              WordPress codex documentation for additional
         *              information on the included hooks.
         *
         *              This method should be used to add any filters or actions
         *              that are necessary to the core of your extension only.
         *              Hooks that are relevant to meta boxes, widgets and
         *              the like can be placed in their respective files.
         * 
         * @see Add Action Hook at https://developer.wordpress.org/reference/functions/add_action/
         * @see Add Filter Hook at https://developer.wordpress.org/reference/functions/add_filter/
         */

        /**
         * Include necessary files
         *
         * @access      private
         * @since       1.0.0
         * @return      void
         */
        private function includes() {

            /**
             * @todo    Make Sure to only include the file(s) you need
             *
             */

            // DB Relational Query for Post Type and Post Meta
            require_once NOTE_FOR_POSTS_DIR . 'includes/class-relational-db-query.php';

            // Friendly HTML Output Class
            require_once NOTE_FOR_POSTS_DIR . 'includes/class-friendly-html.php';

            // Custom Widget with Options
            require_once NOTE_FOR_POSTS_DIR . 'includes/widget-register.php';

            // Register Custom Post Type
            require_once NOTE_FOR_POSTS_DIR . 'includes/custom-post-type.php';

            // Register Custom Taxonomy for Custom Post Type
            require_once NOTE_FOR_POSTS_DIR . 'includes/taxonomy-tags.php';

            // Load Text Domain
            require_once NOTE_FOR_POSTS_DIR . 'includes/text-domain.php';

            // Short Code Functions
            require_once NOTE_FOR_POSTS_DIR . 'includes/shortcodes.php';

            // Frontend Includes
            if ( !is_admin() ) {

                // Enqueue Scripts in the Public
                require_once NOTE_FOR_POSTS_DIR . 'includes/enqueue-scripts.php';  

                // Inline Footer Scripts in the Public
                require_once NOTE_FOR_POSTS_DIR . 'includes/wp-footer-scripts.php';           
            }

            // Admin Includes
            if ( is_admin() || ( defined( 'WP_CLI' ) && WP_CLI ) ) {

                // Manage Custom Columns
                require_once NOTE_FOR_POSTS_DIR . 'includes/admin/manage-columns.php';

                // Enqueue Scripts in the Admin
                require_once NOTE_FOR_POSTS_DIR . 'includes/admin/admin-enqueue-scripts.php';

                 // Footer scripts in the Admin
                require_once NOTE_FOR_POSTS_DIR . 'includes/admin/admin-footer-options.php';

                // Register Custom Meta Box for Custom Post Type
                require_once NOTE_FOR_POSTS_DIR . 'includes/admin/metabox-save-post.php';

                 // Manage Custom Columns
                require_once NOTE_FOR_POSTS_DIR . 'includes/admin/manage-columns.php';

                // WP AJAX Request
                require_once NOTE_FOR_POSTS_DIR . 'includes/admin/wp-ajax-action.php';
              
                // Sub Menu for Settings 
                require_once NOTE_FOR_POSTS_DIR . 'includes/admin/sub-menu-options.php';
                
             }
        }


        /**
         * The name of the plugin used to uniquely identify it within the context of
         * WordPress and to define internationalization functionality.
         *
         * @since     1.0.0
         * @return    string    The name of the plugin.
         */
        public function get_plugin_name() {
            return $this->plugin_name;
        }


    }
} // End if class_exists check


function N4P() {
    return Note_For_Posts::instance();
}
// Run the Plugin Instance
N4P();

